main.go

```diff
@@ -133,6 +143,24 @@ func main() {
 
 }
 
+func chainMiddleware(handler http.Handler, middlewares ...middleware) http.Handler {
+	for _, m := range middlewares {
+		handler = m(handler)
+	}
+	return handler
+}
+
+func requestLoggingMiddleWare(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+
+		//Defer in case one of the handlers down the line calls panic
+		defer func() { log.Println(r.Method, r.URL.Path, time.Since(start)) }()
```

Comment:
Does defer still guarantee order? so if you log the same thing several times is this defer running on the same thread or something so the logs always come out the same?

Reply:
defer runs on the same thread.

Reply:
Yeah defer runs on the same thread. Its sort of like a finally block in Java